File: app/src/main/java/com/whatchapp/hourlybuzz/Schedule.kt
package com.whatchapp.hourlybuzz

import java.time.Instant
import java.time.ZoneId

/** Clock maths. Pure functions, unit-tested in ScheduleTest. */
object Schedule {

    val tickMs: Long
        get() = if (Config.TEST_MODE) Config.TEST_VIBRATE_SECONDS * 1000L
        else Config.VIBRATE_EVERY_MINUTES * 60_000L

    val cardMs: Long
        get() = if (Config.TEST_MODE) Config.TEST_CARD_SECONDS * 1000L
        else Config.CARD_EVERY_MINUTES * 60_000L

    /** Start of the local day containing [now]. */
    fun localMidnight(now: Long, zone: ZoneId = ZoneId.systemDefault()): Long =
        Instant.ofEpochMilli(now).atZone(zone).toLocalDate()
            .atStartOfDay(zone).toInstant().toEpochMilli()

    /**
     * Next time strictly after [now] that is a whole multiple of [periodMs]
     * counted from local midnight: :00, :05, ... for 5 minutes; the top of the
     * next hour for 60 minutes.
     */
    fun nextAligned(now: Long, periodMs: Long, zone: ZoneId = ZoneId.systemDefault()): Long {
        val start = localMidnight(now, zone)
        return start + ((now - start) / periodMs + 1) * periodMs
    }

    /** Most recent aligned time at or before [now]. */
    fun prevAligned(now: Long, periodMs: Long, zone: ZoneId = ZoneId.systemDefault()): Long {
        val start = localMidnight(now, zone)
        return start + ((now - start) / periodMs) * periodMs
    }

    /** True when [time] is exactly on a card boundary. */
    fun isCardTime(time: Long, zone: ZoneId = ZoneId.systemDefault()): Boolean =
        prevAligned(time, cardMs, zone) == time

    /** Number of whole card periods since local midnight (the hour, for hourly cards). */
    fun cardSlot(time: Long, zone: ZoneId = ZoneId.systemDefault()): Long =
        (time - localMidnight(time, zone)) / cardMs

    /** Local hour of day (0-23). */
    fun hourOf(time: Long, zone: ZoneId = ZoneId.systemDefault()): Int =
        Instant.ofEpochMilli(time).atZone(zone).hour

    /** Index wrapped into [0, size), also for negative numbers. */
    fun wrapIndex(index: Int, size: Int): Int =
        if (size <= 0) 0 else ((index % size) + size) % size

    /** Milliseconds as "m:ss" (or "h:mm:ss" for an hour or more). */
    fun formatCountdown(ms: Long): String {
        val total = maxOf(0L, (ms + 999) / 1000)
        val h = total / 3600
        val m = (total % 3600) / 60
        val s = total % 60
        return if (h > 0) "%d:%02d:%02d".format(h, m, s) else "%d:%02d".format(m, s)
    }
}

File: js/schedule.js
/*
 * Pure scheduling helpers. No Tizen or DOM dependencies, so they can be
 * unit-tested with Node (see test/schedule.test.js).
 */
(function (root) {
  "use strict";

  // Start of the local day containing `now` (ms since epoch).
  function localMidnight(now) {
    var d = new Date(now);
    return new Date(d.getFullYear(), d.getMonth(), d.getDate()).getTime();
  }

  // Next time strictly after `now` that is a whole multiple of `periodMs`
  // counted from local midnight. With a 5 minute period that is :00, :05, ...;
  // with 60 minutes it is the top of the next hour.
  function nextAligned(now, periodMs) {
    var start = localMidnight(now);
    var steps = Math.floor((now - start) / periodMs) + 1;
    return start + steps * periodMs;
  }

  // Most recent aligned time at or before `now`.
  function prevAligned(now, periodMs) {
    var start = localMidnight(now);
    return start + Math.floor((now - start) / periodMs) * periodMs;
  }

  // Index wrapped into [0, length), also for negative numbers.
  function wrapIndex(index, length) {
    if (!length) return 0;
    return ((index % length) + length) % length;
  }

  // True when an image boundary was missed (e.g. the app was killed and
  // relaunched by the backup alarm) and should still be shown.
  function missedImage(now, periodMs, lastImageAt, graceMs) {
    var due = prevAligned(now, periodMs);
    return lastImageAt < due && now - due <= graceMs;
  }

  // Milliseconds as "m:ss" (or "h:mm:ss" for an hour or more).
  function formatCountdown(ms) {
    var total = Math.max(0, Math.ceil(ms / 1000));
    var h = Math.floor(total / 3600);
    var m = Math.floor((total % 3600) / 60);
    var s = total % 60;
    var ss = (s < 10 ? "0" : "") + s;
    if (h > 0) return h + ":" + (m < 10 ? "0" : "") + m + ":" + ss;
    return m + ":" + ss;
  }

  var Schedule = {
    localMidnight: localMidnight,
    nextAligned: nextAligned,
    prevAligned: prevAligned,
    wrapIndex: wrapIndex,
    missedImage: missedImage,
    formatCountdown: formatCountdown
  };

  if (typeof module !== "undefined" && module.exports) {
    module.exports = Schedule;
  } else {
    root.Schedule = Schedule;
  }
})(this);
